presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/LogFilter.java

```diff
@@ -0,0 +1,288 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.teradata.prestomanager.agent;
+
+import com.google.common.collect.ImmutableMap;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayDeque;
+import java.util.ConcurrentModificationException;
+import java.util.Deque;
+import java.util.StringJoiner;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Class for filtering the contents of log files
+ *
+ * Uses a regular expression to match log entries, and treats
+ * non-matching lines as part of the entry on the previous line.
+ *
+ * @see Builder
+ */
+public class LogFilter
+{
+    private final Path file;
+    private final Pattern logPattern;
+    private final String lineSeparator;
+    private final ImmutableMap<String, Predicate<String>> namedGroupFilters;
+    private final ImmutableMap<Integer, Predicate<String>> numberedGroupFilters;
+    private final Integer maxEntries;
+    private final boolean keepFirst;
+
+    private LogFilter(Path file, Pattern logPattern, String lineSeparator,
+            ImmutableMap<String, Predicate<String>> namedGroupFilters,
+            ImmutableMap<Integer, Predicate<String>> numberedGroupFilters,
+            Integer maxEntries, boolean keepFirst)
+            throws FileNotFoundException
+    {
+        this.file = requireNonNull(file);
+        this.logPattern = requireNonNull(logPattern);
+        this.lineSeparator = requireNonNull(lineSeparator);
+        this.namedGroupFilters = requireNonNull(namedGroupFilters);
+        this.numberedGroupFilters = requireNonNull(numberedGroupFilters);
+        this.maxEntries = maxEntries; // null -> no maximum
+        this.keepFirst = keepFirst;
+        if (!Files.exists(file)) {
+            throw new FileNotFoundException(file.toString());
+        }
+    }
+
+    public Stream<String> streamEntries()
+            throws IOException, ConcurrentModificationException
+    {
+        LogEntries logEntryBuilders = Files.lines(file).collect(
+                LogEntries::new,
+                LogEntries::addLine,
+                (a, b) -> {
+                    throw new ConcurrentModificationException();
+                });
+        return logEntryBuilders.streamEntries().map(StringJoiner::toString);
+    }
+
+    public static Builder builder()
+    {
+        return new Builder();
+    }
+
+    /**
+     * A builder for creating {@link LogFilter LogFilters}.
+     *
+     * Builder instances can be safely reused.
```

Comment:
Did you mean it's thread safe or reentrant?
If so, please annotate the class appropriately, e.g. with `@javax.annotation.concurrent.ThreadSafe`.